### PR TITLE
JDF-148 showing how to create new beans using DeltaSpike utilities

### PR DESCRIPTION
**Pull request by rafabene**

### CODE REVIEW

File: deltaspike-beanbuilder/README.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+Technologies: CDI, DeltaSpike
+Summary: Shows how to create new beans using DeltaSpike utilities.
+Target Product: WFK
+
+What is it?
+-----------
+
+This project demonstrates a CDI Portable Extension that uses DeltaSpike utilities to create new Beans.
+
+This extension permits the injection of JPA Entities by its Id without the need of query it. To do that, the extension observes the `ProcessInjectionTarget` event and get all values that needs to be queried. Then in `AfterBeanDiscovery` event, the extension create `Bean` instances using the `BeanBuilder` utility.
```
Comment (pmuir) [inline deltaspike-beanbuilder/README.md:14]: s/its/their/

File: deltaspike-beanbuilder/README.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+Target Product: WFK
+
+What is it?
+-----------
+
+This project demonstrates a CDI Portable Extension that uses DeltaSpike utilities to create new Beans.
+
+This extension permits the injection of JPA Entities by its Id without the need of query it. To do that, the extension observes the `ProcessInjectionTarget` event and get all values that needs to be queried. Then in `AfterBeanDiscovery` event, the extension create `Bean` instances using the `BeanBuilder` utility. 
+
+The project contains very simple JPA Entity class, an extension class, the service registration file for that extension and an Arquillian test to verify the extension is working correctly.
```
Comment (pmuir) [inline deltaspike-beanbuilder/README.md:16]: s/contains/contains a/

File: deltaspike-beanbuilder/pom.xml [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+
+        <!-- Import the EJB API, we use provided scope as the API is included 
+            in JBoss AS 7 -->
+        <dependency>
+            <groupId>org.jboss.spec.javax.ejb</groupId>
+            <artifactId>jboss-ejb-api_3.1_spec</artifactId>
+            <scope>provided</scope>
+        </dependency>
+
+        <!-- Deltaspike API. we use compile scope as we need its API -->
```
Comment (pmuir) [inline deltaspike-beanbuilder/pom.xml:133]: s/ we use compile scope as we need its API/We use compile scope as we need the API to compile against/

File: deltaspike-beanbuilder/pom.xml [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        <dependency>
+            <groupId>org.apache.deltaspike.core</groupId>
+            <artifactId>deltaspike-core-api</artifactId>
+            <scope>compile</scope>
+        </dependency>
+
+        <!-- Deltaspike Impl. we use runtime scope as we its implementation 
+            dependencies only on runtime -->
```
Comment (pmuir) [inline deltaspike-beanbuilder/pom.xml:141]: s/we use runtime scope as we its implementation dependencies only on runtime/We use runtime scope as we only need the implementation classes at runtime/

File: deltaspike-beanbuilder/src/main/java/org/jboss/as/quickstart/deltaspike/beanbuilder/ByIdExtension.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    private List<String> ids = new LinkedList<String>();
+
+    /**
+     * This method is fired for every component class supporting injection that may be instantiated by the container at runtime.
+     * 
+     * It will look for all injection target and collect all nicks that were used and needs to be found on {@link EntityManager}
+     * 
+     * @param pit
+     */
+    public <X extends Object> void processInjectionTarget(@Observes ProcessInjectionTarget<X> pit) {
```
Comment (pmuir) [inline deltaspike-beanbuilder/src/main/java/org/jboss/as/quickstart/deltaspike/beanbuilder/ByIdExtension.java:61]: Shouldn't you only observe ProcessInjectionTarget<Person>?

File: deltaspike-beanbuilder/src/main/java/org/jboss/as/quickstart/deltaspike/beanbuilder/ByIdExtension.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+
+
+    /**
+     * This class represents the {@link ById} annotation with its value. It is used by the {@link BeanBuilder} to set the
+     * {@link Bean} {@link Qualifier}
+     * 
+     */
+    public static class ByIdLiteral extends AnnotationLiteral<ById> implements ById {
```
Comment (pmuir) [inline deltaspike-beanbuilder/src/main/java/org/jboss/as/quickstart/deltaspike/beanbuilder/ByIdExtension.java:97]: I would make these top level classes.

File: deltaspike-beanbuilder/src/main/java/org/jboss/as/quickstart/deltaspike/beanbuilder/ByIdExtension.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+        private static final long serialVersionUID = 1L;
+
+        private String value;
```
Comment (pmuir) [inline deltaspike-beanbuilder/src/main/java/org/jboss/as/quickstart/deltaspike/beanbuilder/ByIdExtension.java:101]: make this final

File: deltaspike-beanbuilder/src/main/java/org/jboss/as/quickstart/deltaspike/beanbuilder/ByIdExtension.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+
+    /**
+     * This class represents the {@link ById} annotation with its value. It is used by the {@link BeanBuilder} to set the
+     * {@link Bean} {@link Qualifier}
+     * 
+     */
+    public static class ByIdLiteral extends AnnotationLiteral<ById> implements ById {
+
+        private static final long serialVersionUID = 1L;
```
Comment (pmuir) [inline deltaspike-beanbuilder/src/main/java/org/jboss/as/quickstart/deltaspike/beanbuilder/ByIdExtension.java:99]: Generate a serialVersionUID

File: deltaspike-beanbuilder/src/main/java/org/jboss/as/quickstart/deltaspike/beanbuilder/ByIdExtension.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+    /**
+     * This is the {@link ContextualLifecycle} that is used to by the {@link Bean} to create instances of {@link Person}.
+     * 
+     * It uses the {@link EntityManager#find(Class, Object)} to query the {@link Person} from Database
+     * 
+     */
+    public static class PersonContextualLifecycle implements ContextualLifecycle<Person> {
+
+        private String idValue;
```
Comment (pmuir) [inline deltaspike-beanbuilder/src/main/java/org/jboss/as/quickstart/deltaspike/beanbuilder/ByIdExtension.java:122]: make this final. Make the class top level.

File: deltaspike-beanbuilder/src/main/java/org/jboss/as/quickstart/deltaspike/beanbuilder/util/Resources.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    @Produces
+    public Logger produceLog(InjectionPoint injectionPoint) {
+        return Logger.getLogger(injectionPoint.getMember().getDeclaringClass().getName());
+    }
+
+    @Produces
+    @RequestScoped
+    public FacesContext produceFacesContext() {
+        return FacesContext.getCurrentInstance();
+    }
```
Comment (pmuir) [inline deltaspike-beanbuilder/src/main/java/org/jboss/as/quickstart/deltaspike/beanbuilder/util/Resources.java:55]: Do you use the logger or the facesContext?

**Comment by pmuir** [comment]
Merged with fixes :-)

**Comment by rafabene** [comment]
thx

**Pull request closed without merging**